File: sdks/dotnet/samples/DemoApi/Controllers/OrdersController.cs
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Liaiso.AspNetCore.Discovery;

namespace DemoApi.Controllers;

public sealed record AddNoteRequest([Required][property: Description("Note text")] string Text);

public sealed record OrderResponse(
    [Required] int Id,
    [Required] string Item,
    [Required] int Quantity,
    [Required] string Owner);

/// <remarks>
/// A whole-object query binding, so the sample exercises <c>Query.Grouping</c> end to end. The
/// action echoes the DTO the model binder produced, which is the only direct evidence that the
/// key the composer wrote is the key the backend actually read.
/// </remarks>
public sealed class OrderFilter
{
    [Description("Order owner")]
    public string? Owner { get; set; }

    [Description("Smallest quantity to return")]
    public int? MinQuantity { get; set; }

    [Description("Item names to match")]
    public List<string>? Items { get; set; }
}

public sealed record CreateOrderRequest(
    [Required, MinLength(1)][property: Description("Item name")] string Item,
    [Range(1, 100)][property: Description("Quantity")] int Quantity);

[ApiController]
[McpTool]
public sealed class OrdersController : ControllerBase
{
    private static readonly Dictionary<int, object> Orders = new()
    {
        [1] = new { id = 1, item = "mechanical keyboard", quantity = 2, owner = "alice" },
        [2] = new { id = 2, item = "usb-c dock", quantity = 1, owner = "bob" },
    };

    private static readonly Dictionary<int, List<string>> Notes = [];

    [HttpGet("/ping")]
    [AllowAnonymous]
    [Description("Health check; requires no identity.")]
    public IActionResult Ping() => Ok(new { pong = true });

    [HttpGet("/me")]
    [Authorize]
    [Description("Returns the caller's identity.")]
    public IActionResult Me() => Ok(new { name = User.Identity?.Name });

    [HttpGet("/orders/{id:int}")]
    [Authorize(Policy = "OrdersRead")]
    [ProducesResponseType(typeof(OrderResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [Description("Fetches one order by id.")]
    public IActionResult GetOrder([Description("Order id")] int id) =>
        Orders.TryGetValue(id, out var order) ? Ok(order) : NotFound();

    [HttpGet("/orders")]
    [Authorize(Policy = "OrdersRead")]
    [Description("Searches orders by a filter object, echoing the filter the binder produced.")]
    public IActionResult Search(
        [FromQuery] OrderFilter filter,
        [Description("Largest number of orders to return")] int? take)
    {
        object[] matched = [.. Orders.Values
            .Where(order => Match(order, filter))
            .Take(take ?? 10)];
        return Ok(new { echoed = filter, take, matched });
    }

    /// <remarks>
    /// The same binding with an explicit binder name, which the model binder reads as
    /// <c>f.Owner</c>. Flattened, the tool composes <c>?Owner=</c> and the DTO arrives empty;
    /// grouped, it composes <c>?f.Owner=</c> and binds. The echo shows which one happened.
    /// </remarks>
    [HttpGet("/orders/aliased")]
    [Authorize(Policy = "OrdersRead")]
    [Description("Searches orders through an aliased filter object; echoes what bound.")]
    public IActionResult SearchAliased([FromQuery(Name = "f")] OrderFilter filter) =>
        Ok(new { echoed = filter });

    private static bool Match(object order, OrderFilter filter)
    {
        Type shape = order.GetType();
        string owner = (string)shape.GetProperty("owner")!.GetValue(order)!;
        string item = (string)shape.GetProperty("item")!.GetValue(order)!;
        int quantity = (int)shape.GetProperty("quantity")!.GetValue(order)!;
        return (filter.Owner is null || filter.Owner == owner)
            && (filter.MinQuantity is null || quantity >= filter.MinQuantity)
            && (filter.Items is null || filter.Items.Count == 0 || filter.Items.Contains(item));
    }

    [HttpGet("/orders/{id:int}/receipt")]
    [Authorize]
    [Description("Returns an order's receipt; only the order's owner may see it.")]
    public IActionResult GetReceipt([Description("Order id")] int id)
    {
        if (!Orders.TryGetValue(id, out var order))
        {
            return NotFound();
        }
        string? owner = order.GetType().GetProperty("owner")?.GetValue(order) as string;
        return owner == User.Identity?.Name ? Ok(new { id, receipt = $"receipt-{id}" }) : Forbid();
    }

    [HttpGet("/admin/audit")]
    [Authorize(Roles = "admin")]
    [Description("Audit log; admin role only.")]
    public IActionResult Audit() => Ok(new { entries = Notes.Count });

    [HttpGet("/reports/summary")]
    [Authorize(Policy = "BusinessHours")]
    [Description("Daily summary; closed outside business hours.")]
    public IActionResult Summary() => Ok(new { orders = Orders.Count });

    [HttpPost("/orders/{id:int}/notes")]
    [EndpointName("AddOrderNote")]
    [Authorize(Policy = "OrdersRead")]
    [Description("Adds a note to an order.")]
    public IActionResult AddNote(
        [Description("Order id")] int id,
        [FromQuery, Description("Whether to notify the order's owner")] bool notify,
        [FromBody] AddNoteRequest request)
    {
        if (!Orders.ContainsKey(id))
        {
            return NotFound();
        }
        if (!Notes.TryGetValue(id, out List<string>? list))
        {
            Notes[id] = list = [];
        }
        list.Add(request.Text);
        return Ok(new { id, notes = list, notified = notify });
    }

    [HttpPost("/orders")]
    [EndpointName("CreateOrder")]
    [Authorize(Policy = "OrdersRead")]
    [Description("Creates a new order.")]
    public IActionResult CreateOrder([FromBody] CreateOrderRequest request)
    {
        int id = Orders.Count == 0 ? 1 : Orders.Keys.Max() + 1;
        var order = new { id, item = request.Item, quantity = request.Quantity, owner = User.Identity?.Name };
        Orders[id] = order;
        return Ok(order);
    }
}

File: sdks/nestjs/samples/demo-api/src/orders.controller.ts
import {
  Body,
  Controller,
  ForbiddenException,
  Get,
  NotFoundException,
  Param,
  ParseBoolPipe,
  ParseIntPipe,
  Post,
  Query,
  Req,
  UseGuards,
} from "@nestjs/common";
import { Transform, Type } from "class-transformer";
import { McpTool } from "@sk-mcp/sdk-nestjs";
import {
  IsArray,
  IsInt,
  IsNotEmpty,
  IsOptional,
  IsString,
  Max,
  Min,
  MinLength,
} from "class-validator";
import {
  AdminRoleGuard,
  BusinessHoursGuard,
  JwtGuard,
  OrdersReadGuard,
  type AuthedRequest,
} from "./auth.js";

export class CreateOrderDto {
  @IsString()
  @IsNotEmpty()
  @MinLength(1)
  item!: string;

  @IsInt()
  @Min(1)
  @Max(100)
  quantity!: number;
}

export class AddNoteDto {
  @IsString()
  @IsNotEmpty()
  text!: string;
}

/**
 * A named `@Query('filter')` binding, so the sample exercises `query.grouping` end to end. The
 * handler echoes the DTO the pipe produced, which is the only direct evidence that the key the
 * composer wrote is the key Express actually parsed.
 */
export class OrderFilterDto {
  @IsString()
  @IsOptional()
  owner?: string;

  @IsInt()
  @IsOptional()
  @Type(() => Number)
  minQuantity?: number;

  /**
   * Guard: a query key that appears once parses to a string, not a one-element array — the same
   * in `?items=a` and `?filter[items]=a`. Every Express host with an array query member needs
   * this, grouped or not.
   */
  @IsArray()
  @IsString({ each: true })
  @IsOptional()
  @Transform(({ value }) =>
    value === undefined || Array.isArray(value) ? value : [value],
  )
  items?: string[];
}

export class OrderResponse {
  @IsInt()
  id!: number;

  @IsString()
  item!: string;

  @IsInt()
  quantity!: number;

  @IsString()
  owner!: string;
}

interface Order {
  readonly id: number;
  readonly item: string;
  readonly quantity: number;
  readonly owner: string;
  readonly notes: string[];
}

@Controller()
@McpTool()
export class OrdersController {
  private readonly orders = new Map<number, Order>();
  private nextId = 1;

  @Get("ping")
  @McpTool({ description: "Health check; requires no identity." })
  ping(): { pong: boolean } {
    return { pong: true };
  }

  @Get("me")
  @UseGuards(JwtGuard)
  @McpTool({ description: "Returns the caller's identity." })
  me(@Req() request: AuthedRequest): { name: unknown } {
    return { name: request.user?.sub };
  }

  @Get("orders")
  @UseGuards(JwtGuard, OrdersReadGuard)
  @McpTool({
    description:
      "Searches orders by a filter object, echoing the filter the pipe produced.",
  })
  search(@Query("filter") filter: OrderFilterDto): {
    echoed: OrderFilterDto;
    matched: Order[];
  } {
    const wanted = filter ?? {};
    return {
      echoed: wanted,
      matched: [...this.orders.values()].filter(
        (order) =>
          (wanted.owner === undefined || order.owner === wanted.owner) &&
          (wanted.minQuantity === undefined ||
            order.quantity >= wanted.minQuantity) &&
          (wanted.items === undefined ||
            wanted.items.length === 0 ||
            wanted.items.includes(order.item)),
      ),
    };
  }

  @Get("orders/:id")
  @UseGuards(JwtGuard, OrdersReadGuard)
  @McpTool({
    description: "Fetches one order by id.",
    responses: { 200: OrderResponse, 404: {} },
  })
  getOrder(@Param("id", ParseIntPipe) id: number): Order {
    return this.require(id);
  }

  @Get("orders/:id/receipt")
  @UseGuards(JwtGuard)
  @McpTool({
    description:
      "Returns an order's receipt; only the order's owner may see it.",
  })
  getReceipt(
    @Param("id", ParseIntPipe) id: number,
    @Req() request: AuthedRequest,
  ): { id: number; total: number } {
    const order = this.require(id);
    if (order.owner !== request.user?.sub) {
      throw new ForbiddenException();
    }
    return { id: order.id, total: order.quantity };
  }

  @Get("admin/audit")
  @UseGuards(JwtGuard, AdminRoleGuard)
  @McpTool({ description: "Audit log; admin role only." })
  audit(): { entries: number } {
    return { entries: this.orders.size };
  }

  @Get("reports/summary")
  @UseGuards(JwtGuard, BusinessHoursGuard)
  @McpTool({ description: "Daily summary; closed outside business hours." })
  summary(): { orders: number } {
    return { orders: this.orders.size };
  }

  @Post("orders/:id/notes")
  @UseGuards(JwtGuard, OrdersReadGuard)
  @McpTool({ description: "Adds a note to an order." })
  addOrderNote(
    @Param("id", ParseIntPipe) id: number,
    @Query("notify", new ParseBoolPipe({ optional: true }))
    notify: boolean | undefined,
    @Body() note: AddNoteDto,
  ): { id: number; notes: string[]; notified: boolean } {
    const order = this.require(id);
    order.notes.push(note.text);
    return { id: order.id, notes: order.notes, notified: notify === true };
  }

  @Post("orders")
  @UseGuards(JwtGuard, OrdersReadGuard)
  @McpTool({ description: "Creates a new order." })
  createOrder(
    @Body() body: CreateOrderDto,
    @Req() request: AuthedRequest,
  ): Order {
    const order: Order = {
      id: this.nextId++,
      item: body.item,
      quantity: body.quantity,
      owner: String(request.user?.sub ?? "unknown"),
      notes: [],
    };
    this.orders.set(order.id, order);
    return order;
  }

  private require(id: number): Order {
    const order = this.orders.get(id);
    if (order === undefined) {
      throw new NotFoundException(`order ${String(id)} not found`);
    }
    return order;
  }
}
